src/audio/module_adapter/module_adapter.c

```diff
@@ -208,6 +225,118 @@ static int module_adapter_sink_src_prepare(struct comp_dev *dev)
 	return ret;
 }
 
+#if CONFIG_ZEPHYR_DP_SCHEDULER
+static int module_adapter_dp_queue_prepare(struct comp_dev *dev)
+{
+	struct list_item *blist;
+	struct processing_module *mod = comp_get_drvdata(dev);
+	int i;
+	int ret;
+	struct dp_queue *dp_queue;
+	int dp_mode = dev->is_shared ? DP_QUEUE_MODE_SHARED : DP_QUEUE_MODE_LOCAL;
+
+	/* for DP processing we need to create a DP QUEUE for each module input/output
+	 * till pipeline2.0 is ready, DP processing requires double buffering
+	 *
+	 * first, set all parameters by calling "module prepare" with pointers to
+	 * "main" audio_stream buffers
+	 */
+	ret = module_adapter_sink_src_prepare(dev);
+	if (ret)
+		return ret;
+
+	 /*
+	  * second step - create a "shadow" cross-core DpQueue for existing buffers
+	  * and copy stream parameters to shadow buffers
+	  */
+	i = 0;
+	list_for_item(blist, &dev->bsource_list) {
+		struct comp_buffer *source_buffer =
+			container_of(blist, struct comp_buffer, sink_list);
+
+		/* copy IBS & OBS from buffer to be shadowed */
+		size_t ibs = source_get_ibs(audio_stream_get_source(&source_buffer->stream));
+		size_t obs = sink_get_obs(audio_stream_get_sink(&source_buffer->stream));
+
+		/* create a shadow dp queue */
+		dp_queue = dp_queue_create(ibs, obs, dp_mode);
+
+		if (!dp_queue)
+			goto err;
+
+		mod->dp_queue_ll_to_dp[i] = dp_queue;
+		/* it will override source pointers set by module_adapter_sink_src_prepare
+		 * module will use shadow dpQueue for processing
+		 */
+		mod->sources[i] = dp_queue_get_source(dp_queue);
```

Comment:
CONS: lists are 1) slower 2) harder to debug (no direct access to structs in the lists from debugger)

PROS: List are more memory friendly

I don't like lists personally, especially in the "C way", C++ is a different story

But at this point you're right, since there are tables of sinks/sources and there's no need to iterate through the whole list in every cycle - I'll change a table to a list